# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/mediacontrol/MediaControlService.cpp
# Commit: 714b2fb21e23
# Full Hash: 714b2fb21e2341e8e2bf1be1271ca0ed5e526443
# Author: alwu <alwu@mozilla.com>
# Date: 2020-06-09 09:21:34
# Regressor Bug: 1640998
# File Overlap Count: 1
# Description:
#   Bug 1640998 - part5 : set supported keys to the event source r=chunmin
#   
#   This patch will
#   - add a method `SetSupportedMediaKeys()` on `MediaControlKeysEventSource`
#   - set main controller's supported key to the event source
# ==============================================================================

diff -r a48b6554ba9c -r 714b2fb21e23 dom/media/mediacontrol/MediaControlService.cpp
--- a/dom/media/mediacontrol/MediaControlService.cpp	Mon Jun 08 22:16:46 2020 +0000
+++ b/dom/media/mediacontrol/MediaControlService.cpp	Tue Jun 09 02:23:35 2020 +0000
@@ -330,19 +330,21 @@
   MOZ_ASSERT(NS_IsMainThread());
   mMainController = aController;
 
-  // As main controller has been changed, we should disconnect the listener from
-  // the previous controller and reconnect it to the new controller.
-  DisconnectMainControllerEvents();
-
   if (!mMainController) {
     LOG_MAINCONTROLLER_INFO("Clear main controller");
     mSource->SetPlaybackState(MediaSessionPlaybackState::None);
     mSource->SetMediaMetadata(MediaMetadataBase::EmptyData());
+    mSource->SetSupportedMediaKeys(MediaKeysArray());
+    DisconnectMainControllerEvents();
   } else {
     LOG_MAINCONTROLLER_INFO("Set controller %" PRId64 " as main controller",
                             mMainController->Id());
-    ConnectToMainControllerEvents();
+    mSource->SetPlaybackState(mMainController->GetState());
+    mSource->SetMediaMetadata(mMainController->GetCurrentMediaMetadata());
+    mSource->SetSupportedMediaKeys(mMainController->GetSupportedMediaKeys());
+    ConnectMainControllerEvents();
   }
+
   if (StaticPrefs::media_mediacontrol_testingevents_enabled()) {
     if (nsCOMPtr<nsIObserverService> obs = services::GetObserverService()) {
       obs->NotifyObservers(nullptr, "main-media-controller-changed", nullptr);
@@ -350,20 +352,26 @@
   }
 }
 
-void MediaControlService::ControllerManager::ConnectToMainControllerEvents() {
-  MOZ_ASSERT(mMainController);
-  // Listen to main controller's event in order to get its metadata update.
+void MediaControlService::ControllerManager::ConnectMainControllerEvents() {
+  // As main controller has been changed, we should disconnect listeners from
+  // the previous controller and reconnect them to the new controller.
+  DisconnectMainControllerEvents();
+  // Listen to main controller's event in order to propagate the content that
+  // might be displayed on the virtual control interface created by the source.
   mMetadataChangedListener = mMainController->MetadataChangedEvent().Connect(
       AbstractThread::MainThread(), this,
       &ControllerManager::MainControllerMetadataChanged);
-
-  // Update controller's current status to the event source.
-  mSource->SetPlaybackState(mMainController->GetState());
-  mSource->SetMediaMetadata(mMainController->GetCurrentMediaMetadata());
+  mSupportedKeysChangedListener =
+      mMainController->SupportedKeysChangedEvent().Connect(
+          AbstractThread::MainThread(),
+          [this](const MediaKeysArray& aSupportedKeys) {
+            mSource->SetSupportedMediaKeys(aSupportedKeys);
+          });
 }
 
 void MediaControlService::ControllerManager::DisconnectMainControllerEvents() {
   mMetadataChangedListener.DisconnectIfExists();
+  mSupportedKeysChangedListener.DisconnectIfExists();
 }
 
 MediaController* MediaControlService::ControllerManager::GetMainController()